src/frontend/mod.rs

```diff
@@ -1,2 +1,166 @@
+    use ark_relations::r1cs::{
+        ConstraintSynthesizer, ConstraintSystem, ConstraintSystemRef, SynthesisError,
+    };
+    use core::marker::PhantomData;
+
+    /// CubicFCircuit is a struct that implements the FCircuit trait, for the R1CS example circuit
+    /// from https://www.vitalik.ca/general/2016/12/10/qap.html, which checks `x^3 + x + 5 = y`. It
+    /// has 2 public inputs which are used as the state. `z_i` is used as `x`, and `z_{i+1}` is
+    /// used as `y`, and at the next step, `z_{i+1}` will be assigned to `z_i`, and a new `z+{i+1}`
+    /// will be computted.
+    #[derive(Clone, Copy, Debug)]
+    pub struct CubicFCircuit<F: PrimeField> {
+        _f: PhantomData<F>,
+    }
+    impl<F: PrimeField> FCircuit<F> for CubicFCircuit<F> {
+        type Params = ();
+        fn new(_params: Self::Params) -> Self {
+            Self { _f: PhantomData }
+        }
+        fn step_native(self, z_i: Vec<F>) -> Result<Vec<F>, Error> {
+            Ok(vec![z_i[0] * z_i[0] * z_i[0] + z_i[0] + F::from(5_u32)])
+        }
+        fn generate_step_constraints(
+            self,
+            cs: ConstraintSystemRef<F>,
+            z_i: Vec<FpVar<F>>,
+        ) -> Result<Vec<FpVar<F>>, SynthesisError> {
+            let five = FpVar::<F>::new_constant(cs.clone(), F::from(5u32))?;
+            let z_i = z_i[0].clone();
+
+            Ok(vec![&z_i * &z_i * &z_i + &z_i + &five])
+        }
+    }
+
+    /// CustomFCircuit is a circuit that has the number of constraints specified in the
+    /// `n_constraints` parameter. Note that the generated circuit will have very sparse matrices.
+    #[derive(Clone, Copy, Debug)]
+    pub struct CustomFCircuit<F: PrimeField> {
+        _f: PhantomData<F>,
+        pub n_constraints: usize,
+    }
+    impl<F: PrimeField> FCircuit<F> for CustomFCircuit<F> {
+        type Params = usize;
+
+        fn new(params: Self::Params) -> Self {
+            Self {
+                _f: PhantomData,
+                n_constraints: params,
+            }
+        }
+        fn step_native(self, z_i: Vec<F>) -> Result<Vec<F>, Error> {
+            let mut z_i1 = F::one();
+            for _ in 0..self.n_constraints - 1 {
+                z_i1 *= z_i[0];
+            }
+            Ok(vec![z_i1])
+        }
+        fn generate_step_constraints(
+            self,
+            cs: ConstraintSystemRef<F>,
+            z_i: Vec<FpVar<F>>,
+        ) -> Result<Vec<FpVar<F>>, SynthesisError> {
+            let mut z_i1 = FpVar::<F>::new_witness(cs.clone(), || Ok(F::one()))?;
+            for _ in 0..self.n_constraints - 1 {
+                z_i1 *= z_i[0].clone();
+            }
+
+            Ok(vec![z_i1])
+        }
+    }
+
+    /// WrapperCircuit is a circuit that wraps any circuit that implements the FCircuit trait. This
+    /// is used to test the `FCircuit.generate_step_constraints` method.
+    pub struct WrapperCircuit<F: PrimeField, FC: FCircuit<F>> {
```

Comment:
Maybe we can rename to `StepConstraintsTestCircuit` or similar? To avoid adding more names like `WrappingCircuit` which aren't directly tight to their responsability.

Reply:
So, the `WrapperCircuit` in fact is used to take a 'step circuit' (which implements the `FCircuit` trait) and 'convert' it into a 'non-step-circuit' that can be used not to be folded in the folding scheme but in the normal arkworks trait `ConstraintSynthesizer`.
Is a similar 'wrapping' that the `AugmentedFCircuit` does, but without adding all the extra constraints of the `AugmentedFCircuit`, in order to run lighter tests.
I've [added this last sentence in the documentation](https://github.com/privacy-scaling-explorations/folding-schemes/compare/b9bcfc28f66fcb4103ef5965883401fd488b1ac0..2e661e0c3f0abbf654660b84658cef47a4d6982c) of the struct to add more clarity.

Reply:
Thanks for the clarification!